parity/user_defaults.rs

```diff
@@ -14,110 +14,100 @@
+	use serde::de::Error;
+	use journaldb::Algorithm;
 
-	fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-		write!(formatter, "a valid UserDefaults object")
+	pub fn serialize<S>(algorithm: &Algorithm, serializer: S) -> Result<S::Ok, S::Error>
+	where S: Serializer {
+		algorithm.as_str().serialize(serializer)
 	}
 
-	fn visit_map<V>(self, visitor: V) -> Result<Self::Value, V::Error> where V: MapAccess<'a> {
-		let mut map: BTreeMap<String, Value> = Deserialize::deserialize(MapAccessDeserializer::new(visitor))?;
-		let pruning: Value = map.remove("pruning").ok_or_else(|| Error::custom("missing pruning"))?;
-		let pruning = pruning.as_str().ok_or_else(|| Error::custom("invalid pruning value"))?;
-		let pruning = pruning.parse().map_err(|_| Error::custom("invalid pruning method"))?;
-		let tracing: Value = map.remove("tracing").ok_or_else(|| Error::custom("missing tracing"))?;
-		let tracing = tracing.as_bool().ok_or_else(|| Error::custom("invalid tracing value"))?;
-		let fat_db: Value = map.remove("fat_db").unwrap_or_else(|| Value::Bool(false));
-		let fat_db = fat_db.as_bool().ok_or_else(|| Error::custom("invalid fat_db value"))?;
-
-		let mode: Value = map.remove("mode").unwrap_or_else(|| Value::String("active".to_owned()));
-		let mode = match mode.as_str().ok_or_else(|| Error::custom("invalid mode value"))? {
-			"offline" => Mode::Off,
-			"dark" => {
-				let timeout = map.remove("mode.timeout").and_then(|v| v.as_u64()).ok_or_else(|| Error::custom("invalid/missing mode.timeout value"))?;
-				Mode::Dark(Duration::from_secs(timeout))
-			},
-			"passive" => {
-				let timeout = map.remove("mode.timeout").and_then(|v| v.as_u64()).ok_or_else(|| Error::custom("invalid/missing mode.timeout value"))?;
-				let alarm = map.remove("mode.alarm").and_then(|v| v.as_u64()).ok_or_else(|| Error::custom("invalid/missing mode.alarm value"))?;
-				Mode::Passive(Duration::from_secs(timeout), Duration::from_secs(alarm))
-			},
-			"active" => Mode::Active,
-			_ => { return Err(Error::custom("invalid mode value")); },
-		};
-
-		let user_defaults = UserDefaults {
-			is_first_launch: false,
-			pruning: pruning,
-			tracing: tracing,
-			fat_db: fat_db,
-			mode: mode,
-		};
-
-		Ok(user_defaults)
+	pub fn deserialize<'de, D>(deserializer: D) -> Result<Algorithm, D::Error>
+	where D: Deserializer<'de> {
+		let pruning = String::deserialize(deserializer)?;
+		pruning.parse().map_err(|_| Error::custom("invalid pruning method"))
 	}
 }
 
+pub fn mode_deserialize<'de, D>(deserializer: D) -> Result<Mode, D::Error>
+where D: Deserializer<'de> {
+	// this was added for migration purposes. since the format of the `mode`
+	// field in `UserDefaults` changed the deserialization of the whole struct
+	// would fail. this way we keep the existing values for the other fields and
+	// only provide a default for the `mode` field in case it fails to
+	// deserialize.
+	Ok(Mode::deserialize(deserializer).unwrap_or(Mode::Active))
```

Comment:
Maybe we can preserve backwards compatibility like [this](http://play.rust-lang.org/?gist=2a6b710c8a82e83f2aba958bb0172e26&version=stable&mode=debug&edition=2015)?

Reply:
Nice! But apparently `flatten` can only be used with structs and maps:

`2018-07-17 13:11:47  Error loading user defaults file: ErrorImpl { code: Message("can only flatten structs and maps"), line: 1, column: 107 }`

Always prints that when starting although it seems to work. Also the serialization of `Duration` is different so also not backwards compatible.

(I didn't know you could use crates in rust playground that's super useful!)

Reply:
Ah, looks like we're using `serde` `1.0.37` and support for flattening internally tagged enums was only implemented in [`1.0.46`](https://github.com/serde-rs/serde/issues/1189). The difference in `Duration` serialization is also probably due to different versions used in playground in parity codebase.

(rust playground was merged with `Integer32`'s playground implementation [one year ago](https://users.rust-lang.org/t/the-official-rust-playground-now-has-the-top-100-crates-available/11817))